Add dashboard API to get firmware binaries

kuba2k2 opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->
Dashboard API improvements for https://github.com/esphome/dashboard/pull/469. See that PR for description of changes.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

kuba2k2 committed:
Add dashboard API to get firmware binaries

kuba2k2 committed:
Fix wrong variable name

jesserockz reviewed (changes requested):
I dont think the same endpoint should be used for getting the list of available files per platform, it should have its own endpoint

kuba2k2 committed:
Merge remote-tracking branch 'upstream/dev' into feature/dashboard-download

kuba2k2 committed:
Dashboard: move binary downloads list to separate endpoint

kuba2k2 commented:
Okay, should be better now. I don't even know why did I put that in the same endpoint. I'll update the dashboard in a moment.

kuba2k2 committed:
Init colorama in esphome main

kuba2k2 committed:
Dashboard: use Popen() on Windows

kuba2k2 committed:
Fix pylint errors

kuba2k2 commented:
I updated esphome and dashboard to current upstream branch, and updated the list endpoint.

Also, I added a Popen()-based process spawning method, so that the dashboard can be tested on Windows. Nothing changes for Linux/Docker users, it only adds Windows compatibility (which would raise a Python exception prior to that).

jesserockz commented on esphome/__main__.py:
Please make these changes in their own PR

jesserockz commented on esphome/components/esp32/__init__.py:
```diff
@@ -80,6 +80,23 @@ def get_esp32_variant(core_obj=None):
     return (core_obj or CORE).data[KEY_ESP32][KEY_VARIANT]
 
 
+def get_download_types(storage_json=None):
```
This should be required (or it will crash the below code)
```suggestion
def get_download_types(storage_json):
```

jesserockz commented on esphome/components/esp8266/__init__.py:
```diff
@@ -50,6 +50,17 @@ def set_core_data(config):
     return config
 
 
+def get_download_types(storage_json=None):
```
```suggestion
def get_download_types(storage_json):
```

jesserockz commented on esphome/components/rp2040/__init__.py:
```diff
@@ -42,6 +42,17 @@ def set_core_data(config):
     return config
 
 
+def get_download_types(storage_json=None):
```
```suggestion
def get_download_types(storage_json):
```

jesserockz commented on esphome/dashboard/dashboard.py:
```diff
@@ -481,40 +522,61 @@ def post(self):
+            self.send_error(404)
+            return
+
+        file_name = self.get_argument("file", "firmware.bin")
```
Please also check the `type` as a fallback so we can deprecate that param over time

kuba2k2 replied:
Should I also worry about not specifying `type=` at all (and falling back to default `firmware.bin`)? That will not work for RP2040, as `firmware.uf2` is used instead. The type/file parameter should be mandatory, and AFAIK has been passed by the dashboard already at all times.

jesserockz replied:
Making them mandatory is fine. The download endpoint has been documented to be used externally recently for a tasmota ota update to ESPHome so I don't want to break that.

jesserockz commented on esphome/dashboard/dashboard.py:
```diff
@@ -202,7 +203,11 @@ class EsphomeCommandWebSocket(tornado.websocket.WebSocketHandler):
         self._proc = None
+        self._queue = None
         self._is_closed = False
+        # Windows doesn't support non-blocking pipes,
+        # use Popen() with a reading thread instead
+        self._use_popen = os.name == "nt"
```
Please move any dashboard fixes for windows into their own PR

kuba2k2 committed:
Revert Windows fixes

kuba2k2 committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kuba2k2 committed:
Dashboard: fallback to type parameter

kuba2k2 committed:
Merge remote-tracking branch 'upstream/dev' into feature/dashboard-download

kuba2k2 committed:
Dashboard: require either file or type

jesserockz committed:
Merge branch 'dev' into feature/dashboard-download
